### PR TITLE
Add locking around _output StringBuilder to avoid threading issues

### PR DESCRIPTION
**Pull request by roederja2**

### CODE REVIEW

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs [hunk 1]
```diff
@@ -103,7 +103,7 @@ public TestResult(ITest test)
             Test = test;
             ResultState = ResultState.Inconclusive;
 
-#if NETSTANDARD1_4
+#if !PARALLEL
```
[the remark this answers is not in the document]
Comment (roederja2) [reply src/NUnitFramework/framework/Internal/Results/TestResult.cs]: The idea was to enable locking around the OutWriter by the same switch that enables the RwLock around the other properties. I should probably have switched the If around instead of using `!PARALLEL`.

[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitFramework/framework/Internal/Results/TestResult.cs]: The PARALLEL define means "build in parallel support" so it's related to platforms but isn't really the same thing. For example, we used to have a compact framework build. At one time we didn't support parallel on it but later we did. So this is a policy switch that defines NUnit capabilities, not specifically a platform switch.

In the case of Synchronized, we do it on all platforms that have it. .NET 1.4 doesn't have that property. As you have modified it, any platform that doesn't build in parallel execution will try to use Synchronized. Are all platforms building? If so, maybe the conditional compile is obsolete. I'm not up on what we are and aren't building these days, so someone else will need to check.

Bottom line, however, we probably __do__ want the lock on all builds that define PARALLEL. It's just the Synchronized bit that seems in question.

Thing to check is whether all writes will actually use the lock. Depends on whether they write to Output or _output.

[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitFramework/framework/Internal/Results/TestResult.cs]: OK. That makes sense, at least until / unless we want to support parallel tests on a platform without a synchronized text writer. That seems unlikely.

[the remark this answers is not in the document]
Comment (roederja2) [reply src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Thinking about this again: As I'm not 100% sure about where exactly the threading issue comes from I'm thinking we might have to back port the synchronized TextWriter back to NetStandard1.4 - or is it guaranteed that there is no concurrent access to OutWriter on that platform (reading or writing).

**Comment by roederja2** [comment]
Thanks for merging. I'm not convinced the behaviour is right on netstandard1.4 though - while it's not different than before my change I don't understand why OutWriter doesn't need to be synchronised on that platform.

**Comment by oznetmaster** [comment]
This seemingly is depending on an internal implementation specific implementation of the synchronized TextWriter.  There is no guarantee that synchronized TextWriter will be implemented identically on all platforms.

It seems like another way of solving this problem needs to be found.

As a side note, it is usually recommended that locks not be taken on a public property, but on an object underlying the property, in this case _outWriter.

**Comment by roederja2** [comment]
I agree with all you said. However, my fix works on the MSFT implementation of it and it won't make it worse if it's implemented differently. The only way I see to fix this "properly" is to roll your own synchronised TextWriter that exposes the synchronisation object explicitly. That has the advantage that it also works on NetStandard1.4 . I have an implementation of this almost ready.

**Comment by jnm2** [comment]
Would it be possible to write a test that repros this?

**Comment by roederja2** [comment]
Repros what exactly? The threading issue? Certainly not in a reliable way.

**Comment by jnm2** [comment]
@roederja2 Hopefully I'm not getting mixed up with a work project, but I think we have a few tests that try (and succeed) at causing a race condition within a half a second of multi-threaded contention. It's hard but if we succeed in writing such a test, it'll make sure future implementations don't inadvertently break your scenario.
